### ISSUE TITLE
ts-mockito support

### ISSUE THREAD

**Issue by shaungrady** [getsaf/shallow-render#146]
I would love to have support for [`ts-mockito`](https://github.com/NagRock/ts-mockito)…

Problem:

```ts
// Spec
import { instance, mock, when } from 'ts-mockito';

const mockedActivatedRoute = mock(ActivatedRoute);
when(mockedActivatedRoute.outlet).thenReturn('example');

let shallow: Shallow<TestComponent>;

shallow = new Shallow(TestComponnet, TestModule)
  .mock(ActivatedRoute, instance(mockedActivatedRoute))
```

```ts
// Component
export class TestComponent {
  constructor(private activatedRoute: ActivatedRoute) {
    console.log(activatedRoute.outlet) // => undefined
  }
}
```

… which looks to be caused by the spreading happening here:
https://github.com/getsaf/shallow-render/blob/master/lib/shallow.ts#L111

Maybe some method or flag for providing mocks that won't be adulterated by `shallow-render`?

**Comment by getsaf** [comment]
Thanks for the submission, this is something I've considered trying to address before, but haven't found a fool-proof way to solve without an explicit call to a function other than `mock`.

TL;DR;
You should be able to do this with `provideMock`. It was designed for just this use case.
```typescript
shallow = new Shallow(TestComponnet, TestModule)
  .provideMock({provider: ActivatedRoute, useValue: instance(mockedActivatedRoute)});
```

Which is is effectively:
```typescript
shallow = new Shallow(TestComponnet, TestModule)
  .provide({provider: ActivatedRoute, useValue: instance(mockedActivatedRoute)})
  .dontMock(ActivatedRoute);
```
---------

Here's my thought process when trying to solve it directly in the `mock` function. Sorry if it seems like rambling, I've really just taken this opportunity to think it all through and write it down here so I don't forget later :-)

The goal would really be to allow passing in a mock that is any class instance and have the provided mock maintain the instance through the test process.

I considered a check to see if the mock is a plain object with something like:
```typescript
  mock<TMock>(mockClass: Type<TMock> | InjectionToken<TMock>, stubs: RecursivePartial<TMock>): this {
    const mock = this.setup.mocks.get(mockClass) || {};
    if (mock.constructor === Object) {
      this.setup.mocks.set(mockClass, {...mock, ...stubs as object});
    } else {
      this.setup.mocks.set(
        mockClass,
        // v-- This mutates the original input object when mock is called multiple times during the test
        // process (which is common)
        Object.assign(mock, stubs)
      );
    }
    return this;
  }
```

The negative side-effect is that now, it's now possible for Shallow to mutate the mock object that is sent in which could be a hard-to-find problem if you run into it during your testing.

For example:
Currently, shallow will never mutate anything you pass into it. It will only create copies along the way.

Consider this common pattern where a prepackaged mock is exported from a test helper:
#### MockTheme.ts
```typescript
import { AppTheme } from '~/common/app-theme';

// Provide a common mocked theme object for testing
export const mockTheme = new AppTheme({bg: 'red', fg: 'blue'});
```

> Note, I definitely don't condone this kind of mocking, it's problematic for many reasons, but I also don't intend for Shallow to have opinions about this strategy or be incompatible with it.


```typescript
import { mockTheme } from '~/testHelpers/mockTheme';

describe('the nuances of mutating test mocks in shallow', () => {
  let shallow: Shallow<MyComponent>;

  beforeEach(() => {
    shallow = new Shallow(MyComponent, MyModule)
      .mock(AppTheme, mockTheme);
  });

  it('does not mutate mockFoo when further mocks are applied', async () => {
    const {find} = await shallow
      // Individual tests are allowed to provide additional mock
      // properties on top of existing mocks
      .mock(AppTheme, {fg: 'pink'})
      .render();
    // Currently, shallow will NOT modify the mockTheme instance so this is safe, but if
    // we allowed mutations on .mock, then the global mockTheme object would be
    // permanently modified so that all future tests would receive `fg: pink`.

    expect(find('.my-text').nativeElement.className).toInclude('pink');
  });
});
```
